src/block/executor.rs

```diff
@@ -0,0 +1,437 @@
+    ///
+    /// # Arguments
+    /// * `inner` - The block device backend.
+    /// * `features` - The features that were negotiated between driver and device.
+    pub fn new(mut inner: B, features: u64) -> Result<Self> {
+        let disk_size = inner.seek(SeekFrom::End(0)).map_err(Error::Seek)?;
+        // This check makes sense only if VIRTIO_BLK_F_BLK_SIZE feature is
+        // unsupported, which might be okay to assume for now.
+        // If we add support for this feature as well, this will have to be
+        // restricted to -> `if disk_size % blk_size != 0`.
+        if disk_size % SECTOR_SIZE != 0 {
+            warn!(
+                "Disk size {} is not a multiple of sector size {}; \
+                 the remainder will not be visible to the guest.",
+                disk_size, SECTOR_SIZE
+            );
+        }
+
+        Ok(Self {
+            inner,
+            num_sectors: disk_size >> SECTOR_SHIFT,
+            features,
+        })
+    }
+
+    fn has_feature(&self, feature_pos: u32) -> bool {
+        (self.features & (1u64 << feature_pos)) != 0
+    }
+
+    fn num_sectors(&self) -> u64 {
+        self.num_sectors
+    }
+
+    fn len_to_num_sectors(len: u32) -> u64 {
+        let mut num_sectors = len as u64 / SECTOR_SIZE;
+        if len as u64 % SECTOR_SIZE != 0 {
+            num_sectors += 1;
+        }
+        num_sectors
+    }
+
+    fn check_access(&self, mut sectors_count: u64, sector: u64) -> Result<()> {
+        sectors_count = sectors_count
+            .checked_add(sector)
+            .ok_or(Error::InvalidAccess)?;
+        if sectors_count > self.num_sectors() {
+            return Err(Error::InvalidAccess);
+        }
+        Ok(())
+    }
+
+    /// Executes `request` Request on `B` and `mem` and returns the number of bytes that were
+    /// read from the device.
+    ///
+    /// # Arguments
+    /// * `mem` - A reference to the guest memory.
+    /// * `request` - The request to execute.
+    pub fn execute<M: GuestMemory>(&mut self, mem: &M, request: &Request) -> Result<u32> {
+        let offset = request
+            .sector()
+            .checked_shl(u32::from(SECTOR_SHIFT))
+            .ok_or(Error::InvalidAccess)?;
+        self.inner
+            .seek(SeekFrom::Start(offset))
+            .map_err(Error::Seek)?;
+        let mut bytes_from_dev = 0;
+
+        if self.has_feature(VIRTIO_BLK_F_RO) && request.request_type() != RequestType::In {
+            return Err(Error::ReadOnly);
+        }
+        if !self.has_feature(VIRTIO_BLK_F_FLUSH) && request.request_type() == RequestType::Flush {
+            return Err(Error::Unsupported(VIRTIO_BLK_T_FLUSH));
+        }
+
+        match request.request_type() {
+            RequestType::In => {
+                for (data_addr, data_len) in request.data() {
+                    self.check_access(
+                        StdIoBackend::<B>::len_to_num_sectors(*data_len + bytes_from_dev),
```

Comment:
Is it possible to change `check_access` so the conversion from number of bytes to number of sectors happens within, so we can get rid of `len_to_num_sectors` as a separate method as well?

Reply:
The reason behind this split was the fact that for discard/write zeroes we receive the number of sectors in the request and it seemed cleaner to perform the check directly based on that number for those ops. If it hurts readability much, I can revert to one function.

Reply:
That function is no longer needed, in the meantime I found [this commit](https://github.com/oasis-tcs/virtio-spec/commit/bef3ff7f1e5e81bf9734b2f73f0ef26447358805) which clearly states that data for IN and OUT should be a multiple of 512B, so now I'm returning an error if this condition is not met.